cpp/src/arrow/util/hashing.h

```diff
@@ -764,19 +788,16 @@ class BinaryMemoTable : public MemoTable {
   using HashTableType = HashTable<Payload>;
   using HashTableEntry = typename HashTable<Payload>::Entry;
   HashTableType hash_table_;
-
-  std::vector<int32_t> offsets_;
-  std::string values_;
+  BinaryBuilder binary_builder_;
 
   int32_t null_index_ = kKeyNotFound;
 
   std::pair<const HashTableEntry*, bool> Lookup(hash_t h, const void* data,
                                                 int32_t length) const {
     auto cmp_func = [=](const Payload* payload) {
```

Comment:
Again, could you inline cmp_func and use reference capture for conformance with the rest of the codebase?
```suggestion
    auto cmp_func = [&](const Payload* payload) {
```

Reply:
I agree with Antoine here, in the changed version, only binary_builder is captured via shared_ptr copy.